pkg/k8s/watchers/endpoint.go

```diff
@@ -85,10 +88,80 @@ func (k *K8sWatcher) updateK8sEndpointV1(oldEP, newEP *slim_corev1.Endpoints, sw
 	if option.Config.BGPAnnounceLBIP {
 		k.bgpSpeakerManager.OnUpdateEndpoints(newEP)
 	}
+	k.addKubeAPIServerServiceEPs(newEP)
 	return nil
 }
 
 func (k *K8sWatcher) deleteK8sEndpointV1(ep *slim_corev1.Endpoints, swg *lock.StoppableWaitGroup) error {
 	k.K8sSvcCache.DeleteEndpoints(ep, swg)
 	return nil
 }
+
+func (k *K8sWatcher) handleKubeAPIServerServiceEPChanges(desiredIPs map[string]struct{}) {
+	// We must perform a diff on the ipcache.IdentityMetadata map in order to
+	// figure out which IPs are stale and should be removed, before we inject
+	// new IPs into the ipcache. The reason is because kube-apiserver will
+	// constantly reconcile this specific object, even when it's been deleted;
+	// effectively, this means we can avoid listening for the delete event.
+	// Therefore, any changes to this specific object can be handled in a
+	// "flattened" manner, since the most up-to-date form of it will be an add
+	// or update event. The former is sent when Cilium is syncing with K8s and
+	// the latter is sent anytime after.
+	//
+	// For example:
+	//   * if a backend is removed or updated, then this will be in the form of
+	//     an update event.
+	//   * if the entire object is deleted, then it will quickly be recreated
+	//     and this will be in the form of an add event.
+	ips := ipcache.FilterMetadataByLabels(labels.LabelKubeAPIServer)
+	currentIPs := make(map[string]struct{}, len(ips))
+	for _, v := range ips {
+		currentIPs[v] = struct{}{}
+	}
+
+	toRemove := make(map[string]labels.Labels)
+	for ip := range currentIPs {
+		if _, ok := desiredIPs[ip]; !ok {
+			toRemove[ip] = labels.LabelKubeAPIServer
+		}
+	}
+	ipcache.RemoveAllPrefixesWithLabels(
+		toRemove,
+		source.CustomResource,
+		k.policyRepository.GetSelectorCache(),
+		k.policyManager,
+	)
+
+	for ip := range desiredIPs {
+		ipcache.UpsertMetadata(ip, labels.LabelKubeAPIServer)
+	}
+
+	// Use CustomResource as the source similar to the way the CiliumNode
+	// (pkg/node/manager.Manager) handler does because the ipcache entry needs
+	// to be overwrite-able by this handler and the CiliumNode handler. If we
+	// used Kubernetes as the source, then the ipcache entries inserted (first)
+	// by the CN handler wouldn't be overwrite-able by the entries inserted
+	// from this handler.
+	ipcache.IPIdentityCache.TriggerLabelInjection(
+		source.CustomResource,
+		k.policyRepository.GetSelectorCache(),
+		k.policyManager,
+	)
+}
+
+// TODO(christarazi): Convert to subscriber model along with the corresponding
+// EndpointSlice version.
+func (k *K8sWatcher) addKubeAPIServerServiceEPs(ep *slim_corev1.Endpoints) {
```

Comment:
If I follow correctly, it's because the k8s apiserver will reconcile the service to recreate it anyway, and at that point we do a diff to understand the change.

The explanation is here, I think this was arrived at by runtime experimentation by deploying in clusters with different setups:

https://github.com/cilium/cilium/pull/17823/files/23a7511ea2c9a04a7898510b964b6f667da8185e#diff-ea405d61c943ced44dd56867e4372074f83ed454988135d25a989026ddb45e3dR101-R109

Reply:
Copying my response from Slack on this discussion:

> As I mentioned to Joe in an offline sync, we don't need to hook into deletes of the kubernetes Endpoint object because of the kube-apiserver [reconciler](https://github.com/kubernetes/kubernetes/blob/1367cca8fd67b09606b01c0a9e46cef59aef3424/pkg/controlplane/controller.go#L227) of that Service. This is because any delete to that object will result in it being recreated and would be sent as an add event to Cilium. If something removes any information from the object, it will be re-added (reconciled after a few seconds) which will be sent as a update event.
>
> In all cases, any event (add or update) will have the most up-to-date information.
>
> TLDR: Deletes are essentially useless to hook into it because they are always followed by an add event.